packages/tpetra/core/test/ImportExport/Issue_3968.cpp

```diff
@@ -0,0 +1,206 @@
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ************************************************************************
+// @HEADER
+*/
+
+#include "Teuchos_UnitTestHarness.hpp"
+
+#include "Tpetra_Core.hpp"
+#include "Tpetra_Map.hpp"
+#include "Tpetra_Vector.hpp"
+#include "Tpetra_Export.hpp"
+#include "Teuchos_CommHelpers.hpp"
+
+#include <iostream>
+#include <numeric>
+#include <vector>
+
+// Test for Trilinos GitHub Issue #3968.
+
+namespace { // (anonymous)
+
+  using GST = Tpetra::global_size_t;
+  using LO = Tpetra::Map<>::local_ordinal_type;
+
+  // Try to get a 64-bit GlobalOrdinal type; that should be more
+  // likely to manifest atomic update issues.  If we can't, though,
+  // then a 32-bit type is fine.  long long is guaranteed to be at
+  // least 64 bits, so prefer that over long or unsigned long.
+#if defined(HAVE_TPETRA_INST_INT_LONG_LONG)
+  using GO = long long; // always at least 64 bits
+#elif defined(HAVE_TPETRA_INST_INT_LONG)
+  using GO = long; // may be 64 or 32 bits
+#elif defined(HAVE_TPETRA_INST_INT_UNSIGNED_LONG)
+  using GO = unsigned long; // may be 64 or 32 bits
+#else
+  using GO = Tpetra::Map<>::global_ordinal_type;
+#endif
+
+  using map_type = Tpetra::Map<LO, GO>;
+  using vec_type = Tpetra::Vector<GO, LO, GO>;
+  using export_type = Tpetra::Export<LO, GO>;
+
+  Teuchos::RCP<const map_type>
+  createTargetMap (const Teuchos::RCP<const Teuchos::Comm<int>>& comm)
+  {
+    const GST gblNumInds = static_cast<GST> (comm->getSize ());
+    const GO indexBase = 0;
+    return Teuchos::rcp (new map_type (gblNumInds, indexBase, comm));
+  }
+
+  // As overlapping as possible, but avoid the explicitly "locally
+  // replicated" path, just in case Tpetra optimizes for that (it
+  // doesn't really, but it could).
+  Teuchos::RCP<const map_type>
+  createSourceMap (const Teuchos::RCP<const Teuchos::Comm<int>>& comm)
+  {
+    const GST flag = Teuchos::OrdinalTraits<GST>::invalid ();
+    const LO lclNumInds = static_cast<LO> (comm->getSize ());
+    std::vector<GO> gblInds (lclNumInds);
+    std::iota (std::begin (gblInds), std::end (gblInds), GO (0));
+    const GO indexBase = 0;
+    return Teuchos::rcp (new map_type (flag, gblInds.data (), lclNumInds, indexBase, comm));
+  }
+
+  // "Base value" should be some value, not default, not negative (in
+  // case GO is unsigned, which is not default but possible).  Pick
+  // baseValue() > comm->getSize().
+  constexpr GO baseValue () {
```

Comment:
Maybe add:
TEUCHOS_ASSERT(baseValue > comm->getSize());

Reply:
True -- no need for `constexpr` here.